### ISSUE TITLE
[1.7.7] $cart_presenter->present($cart, true) - wrong totals

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#20564]
### Steps to Reproduce

1. Make product A with price 100 CZK.
2. Make product B with price 20 CZK.
3. Make a cart rule that gives product B as a gift to all orders.
4. Add product A to cart. You will get product B automatically.
5. Add product B to cart normally.
6. Output content of this somewhere on the website.
```
$presenter = new CartPresenter();
$presented_cart = $presenter->present($this->context->cart, $shouldSeparateGifts = true);
```
or change `'cart' => $this->cart_presenter->present($this->context->cart),` to `'cart' => $this->cart_presenter->present($this->context->cart, true),` in classes/controller/FrontController and go to cart.

7. Totals should be:
Subtotal: 140 CZK
Discounts: 20 CZK
Total: 120 CZK

8. Totals are:
Subtotal: 180 CZK
Discounts: 40 CZK
Total: 140 CZK

### Cause
**1. Wrong subtotal** - /src/Core/Cart/CartRow - Subtotal is calculated by `$rowData['cart_quantity']`instead of `$rowData['quantity']`. Cart_quantity holds total quantity of this product in cart (2), quantity holds the current row amount (1).
**2. Wrong discount total** - /src/Core/Cart/CartRuleCalculator - Error near `if ((int) $cartRule->gift_product) {`. The gift gets discounted twice. Why? Because it checks, if the row equals the product in the cart rule. But this condition will be met twice, in both passes, once for the actual gift and once for the normally added product.

I tried to fix it by this PR https://github.com/PrestaShop/PrestaShop/pull/17825, but it broke some tests and I don't have the skills to resolve the issue, nor know if my solution is correct.

### Additional information

* PrestaShop version: 1.7.7 beta release
* PHP version: 7.2

**Comment by sarahdib** [comment]
Hello,

<img width="1215" alt="Capture d’écran 2020-08-13 à 10 53 49" src="https://user-images.githubusercontent.com/3034677/90114910-ad6f5480-dd53-11ea-80e9-a6311b4c84b6.png">


It seems that your issue is fixed. I close this issue feel free to open a new one if you reproduce this again.

Thank you!

**Comment by Hlavtox** [comment]
@sarahdib You did not reproduce it.

**Comment by Hlavtox** [comment]
Ping @matks @eternoendless

**Comment by jf-viguier** [comment]
Why closing it so fast ?

**Comment by eternoendless** [comment]
@jf-viguier it's part of the normal procedure we use to avoid having too many non reproductible issues open.

That said, looks like there was a mistake when trying to reproduce, we're investigating.

**Comment by jf-viguier** [comment]
Don't forget $shouldSeparateGifts to true

**Comment by matks** [comment]
> Why closing it so fast ?

When a bug is reported, the QA team attempts to reproduce it. Here it could not be reproduced.

This does not mean the conversation ends there
- a closed issue can still be discussed
- if new items are added and the bug can be reproduced we can re-open the issue

"The issue is closed" does not mean "this topic is closed" 😄  . Conversation and exploration can continue.

It's simply the only way to keep a backlog manageable. "Open issues" are "things we can act on". Things a developer can pick up and start working on. We cannot act on something we cant reproduce.

There is a lot of issues that are opened but actually do not refer to a bug inside prestashop software. It can come from server configuration, 3rd party theme, 3rd party module, overrides, custom code ... so if we cannot reproduce an issue on a standard environment, unless proved otherwise, we consider it does not come from prestashop software (and we then cannot act on it).

But once again: it does not mean topic is closed. It's just an issue status, that we use to search, filter, classify issues while manageing the backlog.

**Comment by sarahdib** [comment]
@Hlavtox sorry I check with a member of the core team :)

**Comment by matthieu-rolland** [comment]
I managed to reproduce this issue on 1.7.7, using the same values as @Hlavtox :

<img width="418" alt="Capture d’écran 2020-08-13 à 11 38 09" src="https://user-images.githubusercontent.com/1784781/90119164-83b92c00-dd59-11ea-8ae7-4ed97d7fd3f6.png">

**Comment by matthieu-rolland** [comment]
@sarahdib the same issue exists on 1.7.6, it's not a 1.7.7 regression.

**Comment by Hlavtox** [comment]
Any update pls? If anyone fixes this, we will finally have gifts working after several years. Pleaseeeeee 🙏.
@matthieu-rolland @jolelievre

**Comment by jf-viguier** [comment]
I agree with you Hlavtox

**Comment by Hlavtox** [comment]
@PrestaShop/qa-functional Can you please retest? The thing seems fixed on 8.1.2. 🤷‍♂️ 

I am trying to break it with multiple gifts, additional discounts and everything just works. :D

![Snímek obrazovky 2023-10-26 142509](https://github.com/PrestaShop/PrestaShop/assets/6097524/a1d14c02-2df8-4936-b23a-b94098ccd656)

To test it, open `classes/controller/FrontController.php` and change 
`'cart' => $this->cart_presenter->present($cart),`
to
`'cart' => $this->cart_presenter->present($cart, true),`

**Comment by florine2623** [comment]
Indeed, it is fixed on PS 8.1.2 :

<img width="1171" alt="Screenshot 2023-10-27 at 14 55 35" src="https://github.com/PrestaShop/PrestaShop/assets/16019289/e6f539da-ebfc-4da8-98a0-5a7d07957b4b">

<img width="1203" alt="Screenshot 2023-10-27 at 14 57 23" src="https://github.com/PrestaShop/PrestaShop/assets/16019289/44f36e24-b568-4806-9f2b-9049df39ca38">

I'm closing this issue then ^^
